### PR TITLE
Add offline signing support to CLI

### LINKED ISSUE #6872: CLI - Support offline transaction signing
#### Problem

Our CLI tools don't facilitate offline transaction signing.  This is a requirement for paper wallets and airgapped key management solutions.

#### Proposed Solution

TX Signing: Add a new flag `--sign-only` which, when passed, stops after the signing step and prints the signature to `stdout` encoded as base58.

TX submission: Add two new flags `--signature=BASE58_SIG` and `--signer-pubkey=PUBKEY` that, when passed, skip the signing step and instead verify that `BASE58_SIG` is a valid signature of the transaction by `PUBKEY`

### PR DESCRIPTION
**Pull request by jackcmay**
#### Problem

No support for offline signing

#### Summary of Changes

The following only apply to: pay, delegate-stake, and deactivate-stake

add a --sign-only parameter that will not send a transaction but instead print the blockhash used (could be the nonce) as well as all the PUBKEY=SIGNATURE pairs.
add --blockhash parameter which provides the ability to pass a blockhash to be used rather than querying the network
-add --signer PUBKEY=SIGNATURE parameter that can be used to construct a transaction without requiring keypairs. All the pubkey/signature pairs required by the transaction must be specified.

Fixes #6872

### CODE REVIEW

**Comment by jackcmay** [comment]
@mvines @t-nelson First time I've touched the CLI stuff, before I continue on can you give me a quick sanity check that this is heading in the right direction?

File: cli/src/cli.rs [hunk 1]
```diff
@@ -405,6 +406,11 @@ pub fn parse_command(matches: &ArgMatches<'_>) -> Result<CliCommandInfo, Box<dyn
             let timestamp_pubkey = value_of(&matches, "timestamp_pubkey");
             let witnesses = values_of(&matches, "witness");
             let cancelable = matches.is_present("cancelable");
+            let sign_only = if matches.is_present("sign_only") {
```
Comment (garious) [inline cli/src/cli.rs]: How is that different than `let sign_only = value_of(&matches, "sign_only");`?

Comment (jackcmay) [reply cli/src/cli.rs]: Heh, yeah, same :-)

**Comment by t-nelson** [comment]
Seem to be on track by my eye!

**Comment by mvines** [comment]
I wouldn't expect the `--sign-only` argument to accept a nonce.    `--sign-only` is just a flag that says, "output the transaction signature and the blockhash used".    Maybe I want to sign-only a non-nonce-based transaction.

On the flip side maybe I want to send a transaction immediately using a nonce instead of recent blockhash.   

So I think there's a completely separate argument (that we haven't really discussed yet in #6872 or elsewhere) that activates nonce-support in the CLI

**Comment by jackcmay** [comment]
Do we expect transactions singed offline to be submitted quickly enough to render a recent blockhash valid?

**Comment by mvines** [comment]
For the custody solution, no.   But my point is that I don't we need/should bake in the assumption that `--sign-only` equals nonce.

**Comment by jackcmay** [comment]
Sounds good, so we can add another option `--nonce blah`

**Comment by mvines** [comment]
"blah" is not valid base58 :trollface:

File: cli/src/cli.rs [hunk 2]
```diff
@@ -703,12 +730,20 @@ fn process_pay(
...
+    nonce: Option<Hash>,
 ) -> ProcessResult {
     check_unique_pubkeys(
         (&config.keypair.pubkey(), "cli keypair".to_string()),
         (to, "to".to_string()),
     )?;
-    let (blockhash, fee_calculator) = rpc_client.get_recent_blockhash()?;
+
+    let (blockhash, fee_calculator) = if let Some(nonce) = nonce {
+        (nonce, rpc_client.get_recent_blockhash()?.1)
```
Comment (jstarry) [inline cli/src/cli.rs]: If `sign_only` is true, there's a good chance there's no network available right? So, when signing only, you might have to accept a fee from the cli because fee calculator won't be available via RPC

**Comment by mvines** [comment]
Worse PR title ever.  🏅

File: cli/src/cli.rs (hunk 2 above)
Comment (jackcmay) [reply cli/src/cli.rs]: Then you must provide a blockhash, no need deal with fees

**Comment by jackcmay** [comment]
@mvines Can't get more straight to the point with the PR title ;-)

File: cli/src/cli.rs (hunk 2 above)
Comment (jstarry) [reply cli/src/cli.rs]: Got it, no fee needed from cli. But this behaviour needs to change when `sign_only == true` so that we don't short circuit here when there's no network

Comment (jackcmay) [reply cli/src/cli.rs]: If `sign_only == true`, no nonce is provided, and there is no network then that's an error.

Comment (jstarry) [reply cli/src/cli.rs]: It looks like this code will error if there is a nonce and `sign_only == true` but there is no network because you are requesting the fee calculator.

File: sdk/src/transaction.rs [hunk 3]
```diff
@@ -246,6 +255,26 @@ impl Transaction {
             .collect())
     }
 
+    /// Replace all the signatures and pubkeys
+    pub fn replace_signatures(&mut self, signers: &[(Pubkey, Signature)]) -> Result<()> {
+        let num_required_signatures = self.message.header.num_required_signatures as usize;
+        if signers.len() != num_required_signatures
+            || self.signatures.len() != num_required_signatures
+            || self.message.account_keys.len() < num_required_signatures
```
Comment (jstarry) [inline sdk/src/transaction.rs]: These checks seem outside of the scope of this method and could lead to confusing errors to the caller of `replace_signatures`. I think that it's fine to rely on our runtime to reject these errors. Thoughts?

File: sdk/src/transaction.rs [hunk 4]
```diff
@@ -246,6 +255,26 @@ impl Transaction {
             .collect())
     }
 
+    /// Replace all the signatures and pubkeys
+    pub fn replace_signatures(&mut self, signers: &[(Pubkey, Signature)]) -> Result<()> {
```
Comment (jstarry) [inline sdk/src/transaction.rs]: How about moving signature verification here instead of inside a separate method? In order to replace a signature it must be verified first? That way `verify` can't be forgotten?

File: sdk/src/transaction.rs [hunk 5]
```diff
@@ -246,6 +255,26 @@ impl Transaction {
...
+        {
+            return Err(TransactionError::InvalidAccountIndex);
+        }
+
+        signers
+            .iter()
+            .enumerate()
+            .for_each(|(i, (pubkey, signature))| {
+                self.signatures[i] = *signature;
+                self.message.account_keys[i] = *pubkey;
```
Comment (jstarry) [inline sdk/src/transaction.rs]: I don't think account keys should be overridden. They should be checked for equality. Otherwise the compiled instructions could point to the wrong account key.

Comment (jackcmay) [reply sdk/src/transaction.rs]: Where do the account keys come from if the online entity doesn't have the keypairs (only the offline entity does)?  Are you thinking the pubkeys should come from a config file?

Couple of thoughts
- account keys are public, if this is a legitimate concern we should probably make those private
- The account keys used in the instructions themselves don't come from the config file (private keypairs), they come from parameters passed on the command line (there may be exceptions).  
- If the account keys change then the signatures would have to still match and if they do then its what the signer intended anyway.

File: sdk/src/transaction.rs (hunk 3 above)
Comment (jackcmay) [reply sdk/src/transaction.rs]: Having these checked up front rather then after the transaction is submitted would be very helpful.  I was also following the example of `get_signing_keypair_positions`.  Maybe we should generate more descriptive errors?

File: clap-utils/src/input_parsers.rs [hunk 6]
```diff
@@ -172,4 +186,25 @@ mod tests {
 
         fs::remove_file(&outfile).unwrap();
     }
+
+    #[test]
+    fn test_pubkeys_sigs_of() {
+        let key1 = Pubkey::new_rand();
+        let key2 = Pubkey::new_rand();
+        let sig1 = Keypair::new().sign_message(&[0u8]);
```
[the remark this answers is not in the document]
Comment (jackcmay) [reply clap-utils/src/input_parsers.rs]: These aren't the same things

File: sdk/src/transaction.rs (hunk 5 above)
Comment (jstarry) [reply sdk/src/transaction.rs]: Hmm good points.

Could account keys come from just the public keys then and when you pass in signers, you don't attempt to sign the transaction?

> account keys are public, if this is a legitimate concern we should probably make those private

Yeah, I think I agree

> The account keys used in the instructions themselves don't come from the config file (private keypairs), they come from parameters passed on the command line (there may be exceptions).

I'm confused by this. Looks like you are still creating a signed transaction with config here: https://github.com/solana-labs/solana/pull/7104/files#diff-6897771877027eaa520b55b24b6dceafR797. That doesn't look right... Can you help me understand why this makes sense?

> If the account keys change then the signatures would have to still match and if they do then its what the signer intended anyway.

Yeah true, but could lead to confusing errors because it wouldn't be obvious that the signature was failing because of mis-ordered signers right?

File: clap-utils/src/input_parsers.rs [hunk 7]
```diff
@@ -50,6 +51,19 @@ pub fn pubkey_of(matches: &ArgMatches<'_>, name: &str) -> Option<Pubkey> {
     value_of(matches, name).or_else(|| keypair_of(matches, name).map(|keypair| keypair.pubkey()))
 }
 
+// Return pubkey/signature pairs for a string of the form pubkey=signature
+pub fn pubkeys_sigs_of(matches: &ArgMatches<'_>, name: &str) -> Option<Vec<(Pubkey, Signature)>> {
+    matches.values_of(name).map(|xs| {
```
[the remark this answers is not in the document]
Comment (jackcmay) [reply clap-utils/src/input_parsers.rs]: xs is actually `Values`, good suggestion though, I changed these names to make them more descriptive.

File: clap-utils/src/input_parsers.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jstarry) [reply clap-utils/src/input_parsers.rs]: Oh, why not? Isn't it the same `Signature`? Feel free to disregard then :)

File: sdk/src/transaction.rs (hunk 5 above)
Comment (jackcmay) [reply sdk/src/transaction.rs]: We don't really have a way to determine whether signer ordering is correct since we don't have anything to compare to between what the CLI caller passes in and what is in the transaction.

As for why we create a signed transaction, the instructions themselves don't contain the pubkey, just the indices (unless they are placed into the instruction data but that's a whole different issue) and therefore get replaced appropriately.  If we are worried about non-signer keys we could replace all the account keys.  A signed transaction matches the source form and order that the pubkey=signature pairs were provided to the offline user when the original was produced.  I would rather not generate the transaction differently whether its offline or online.  If for some use case we encounter a case where the private key is stored in the instruction data we can deal with that appropriately then.

File: clap-utils/src/input_parsers.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jackcmay) [reply clap-utils/src/input_parsers.rs]: `new` takes a full signature array

File: sdk/src/transaction.rs (hunk 5 above)
Comment (jstarry) [reply sdk/src/transaction.rs]: Ah ok! Makes a lot more sense now, thanks for explaining.

File: sdk/src/transaction.rs (hunk 3 above)
Comment (jstarry) [reply sdk/src/transaction.rs]: Never mind, I see that by doing this you are avoiding a panic from a bad index. Looks fine 👍
